File: src/main/java/net/gtaun/shoebill/common/dialog/TabListDialog.java
package net.gtaun.shoebill.common.dialog;

import net.gtaun.shoebill.constant.DialogStyle;
import net.gtaun.shoebill.object.Player;
import net.gtaun.util.event.EventManager;

/**
 * Created by marvin on 29.11.15.
 * Copyright (c) 2015 Marvin Haschker. All rights reserved.
 */
public class TabListDialog extends ListDialog {

    @SuppressWarnings("unchecked")
    public static abstract class AbstractTabListDialogBuilder
            <DialogType extends TabListDialog, DialogBuilderType extends AbstractTabListDialogBuilder<DialogType, DialogBuilderType>>
            extends AbstractListDialogBuilder<DialogType, DialogBuilderType> {

        protected AbstractTabListDialogBuilder(DialogType dialog) {
            super(dialog);
        }

        public DialogBuilderType header(int index, String title) {
            dialog.setHeader(index, title);
            return (DialogBuilderType) this;
        }
    }

    public static class TabListDialogBuilder extends AbstractTabListDialogBuilder<TabListDialog, TabListDialogBuilder> {
        protected TabListDialogBuilder(Player player, EventManager parentEventManager) {
            super(new TabListDialog(player, parentEventManager));
        }
    }

    public static AbstractTabListDialogBuilder<?, ?> create(Player player, EventManager parentEventManager) {
        return new TabListDialogBuilder(player, parentEventManager);
    }

    public static final int MAX_HEADERS = 4;
    private String[] headers = {"", "", "", ""};

    protected TabListDialog(Player player, EventManager eventManager) {
        super(player, eventManager);
        style = DialogStyle.TABLIST;
    }

    public void setHeader(int index, String title) {
        if(index < 0 || index > 3) throw new IllegalArgumentException("Index must be between 0 and 4");
        headers[index] = title;
    }

    @Override
    public String getItemString() {
        String listStr = "";
        int headerCount = usedHeaderCount();
        if(headerCount > 0) {
            style = DialogStyle.TABLIST_HEADERS;
            for(int i = 0; i < headerCount; i++) {
                if(headers[i].length() < 1) continue;
                listStr += headers[i];
                if(i + 1 < headerCount) listStr += "\t";
                else {
                    listStr += "\n";
                    if(headerCount < MAX_HEADERS)
                        listStr += "\n";
                }
            }
        } else style = DialogStyle.TABLIST;
        displayedItems.clear();
        for(ListDialogItem item : items) {
            listStr += item.getItemText();
            displayedItems.add(item);
        }
        return listStr;
    }

    private int usedHeaderCount() {
        int count = 0;
        for(String header : headers) {
            if(header.length() > 0)
                count += 1;
        }
        return count;
    }
}

File: src/main/java/net/gtaun/shoebill/common/dialog/TabListDialog.kt
package net.gtaun.shoebill.common.dialog

import net.gtaun.shoebill.Shoebill
import net.gtaun.shoebill.common.AllOpen
import net.gtaun.shoebill.constant.DialogStyle
import net.gtaun.util.event.EventManager

/**
 * Created by marvin on 29.11.15.
 * Copyright (c) 2015 Marvin Haschker. All rights reserved.
 */
@AllOpen
class TabListDialog protected constructor(eventManager: EventManager) : ListDialog(eventManager) {

    @AllOpen
    class TabListDialogBuilder(parentEventManager: EventManager) :
            AbstractListDialogBuilder<TabListDialog, TabListDialogBuilder>() {

        fun header(index: Int, headerSupplier: DialogTextSupplier): TabListDialogBuilder {
            dialog.setHeader(index, headerSupplier)
            return this
        }

        fun header(index: Int, header: String): TabListDialogBuilder {
            dialog.setHeader(index, header)
            return this
        }

        init {
            dialog = TabListDialog(parentEventManager)
        }

    }

    val headers = arrayOfNulls<DialogTextSupplier?>(4)

    init {
        style = DialogStyle.TABLIST
    }

    fun setHeader(index: Int, titleSupplier: DialogTextSupplier) {
        if (index < 0 || index > 3) throw IllegalArgumentException("Index must be [0..3]")
        headers[index] = titleSupplier
    }

    fun setHeader(index: Int, title: String) = setHeader(index, DialogTextSupplier { title })

    override val itemString: String
        get() {
            var listStr = ""
            val headerCount = usedHeaderCount
            if (headerCount > 0) {
                style = DialogStyle.TABLIST_HEADERS
                for (i in 0 until headerCount) {
                    val headerText = headers[i]?.get(this) ?: continue
                    if (headerText.isEmpty()) continue

                    listStr += headerText
                    if (i + 1 < headerCount)
                        listStr += "\t"
                    else {
                        listStr += "\n"
                        if (headerCount < MAX_HEADERS)
                            listStr += "\n"
                    }
                }
            } else
                style = DialogStyle.TABLIST
            displayedItems.clear()
            for (item in items) {
                listStr += item.itemText
                displayedItems.add(item)
            }
            return listStr
        }

    val usedHeaderCount: Int
        get() = headers.filterNotNull().size

    companion object {

        @JvmStatic
        @JvmOverloads
        fun create(eventManager: EventManager = Shoebill.get().eventManager) = TabListDialogBuilder(eventManager)

        internal val MAX_HEADERS = 4
    }
}
